legion/engine/physics/mesh_splitter_utils/mesh_half_edge.hpp

```diff
@@ -0,0 +1,268 @@
+
+        std::shared_ptr<MeshHalfEdge> nextEdge = nullptr;
+        std::shared_ptr<MeshHalfEdge> pairingEdge = nullptr;
+        std::shared_ptr<MeshHalfEdge> shadowEdge = nullptr;
+
+        std::weak_ptr<SplittablePolygon> owner;
+
+        //TODO: these bools should probably be stored as an array in SplittablePolygon
+        bool isVisited = false;
+        bool isBoundary = false;
+
+        //std::string debugID = "";
+
+        MeshHalfEdge(const math::vec3& pPosition, const math::vec2& pUVs, std::weak_ptr<SplittablePolygon> pOwner) : position(pPosition), uv(pUVs),owner(pOwner)
+        {
+
+        }
+
+        MeshHalfEdge(const math::vec3& pPosition,const math::vec2& pUVs) : position(pPosition),uv(pUVs)
+        {
+
+        }
+
+        MeshHalfEdge(const math::vec3& pPosition) : position(pPosition)
+        {
+
+        }
+
+        auto GetTriangle() 
+        {
+            return std::make_tuple(shared_from_this(), nextEdge, nextEdge->nextEdge);
+        }
+
+        auto GetShadowTriangle()
+        {
+            return std::make_tuple(shadowEdge, nextEdge->shadowEdge, nextEdge->nextEdge->shadowEdge);
+        }
+
+        void SetPairing(std::shared_ptr<MeshHalfEdge>& newPairing)
+        {
+            pairingEdge = newPairing;
+            newPairing->pairingEdge = shared_from_this();
+        }
+
+        void CloneOnShadowEdge()
+        {
+            shadowEdge = std::make_shared<MeshHalfEdge>(position, uv);
+            shadowEdge->isBoundary = isBoundary;
+        }
+
+        static void ConnectIntoTriangle(std::shared_ptr<MeshHalfEdge> first
+            , std::shared_ptr<MeshHalfEdge> second,
+            std::shared_ptr<MeshHalfEdge> third)
+        {
+            first->nextEdge = second;
+            second->nextEdge = third;
+            third->nextEdge = first;
+        }
+ 
+
+        bool IsSplitByPlane(const math::mat4& transform
+            , const math::vec3& planePosition
+            ,const  math::vec3& planeNormal)
+        {
+            int x = 0;
+            int y = 0;
+
+            auto [currentDistFromPlane, nextDistFromPlane] = GetEdgeDistancesFromPlane(transform, planePosition, planeNormal);
+
+            if (currentDistFromPlane > constants::polygonSplitterEpsilon)
+            {
+                x = 1;
+            }
+            else if (currentDistFromPlane < -constants::polygonSplitterEpsilon)
+            {
+                x = -1;
+            }
+
+
+            if (nextDistFromPlane > constants::polygonSplitterEpsilon)
+            {
+                y = 1;
+            }
+            else if (nextDistFromPlane < -constants::polygonSplitterEpsilon)
+            {
+                y = -1;
+            }
+
+            int edgeSplitState = x * y;
+
+
+            return edgeSplitState < 0;
+
+
+        }
+
+        math::vec3 GetWorldCentroid(const math::mat4& transform) const
+        {
+            return transform * math::vec4((position + nextEdge->position) * 0.5f, 1);
+        }
```

Comment:
Since any matrix can be passed in, I would advice naming the function something like getTransformedCentroid

Reply:
yes but GetWorldCentroid explains it better, i will just comment it

Reply:
Please add the comment